lib/tapioca/compilers/dsl/active_record_relations.rb

```diff
@@ -0,0 +1,498 @@
+              klass.create_constant("Elem", value: "type_member(fixed: #{@constant})")
+
+              create_method(klass, "to_ary", parameters: [], return_type: "T::Array[#{@constant}]")
+              create_association_relation_methods(klass)
+            end
+          end
+
+          sig { void }
+          def create_collection_proxy_class
+            superclass = "::ActiveRecord::Associations::CollectionProxy"
+
+            # The relation subclass includes the generated association relation module
+            model.create_class(@associations_collection_proxy_class_name, superclass: superclass) do |klass|
+              klass.create_include(@common_relation_methods_module_name)
+              klass.create_include(@association_relation_methods_module_name)
+              klass.create_constant("Elem", value: "type_member(fixed: #{@constant})")
+
+              create_method(klass, "to_ary", parameters: [], return_type: "T::Array[#{@constant}]")
+              create_association_relation_methods(klass)
+              create_collection_proxy_methods(klass)
+            end
+          end
+
+          sig { params(klass: Parlour::RbiGenerator::ClassNamespace).void }
+          def create_association_relation_methods(klass)
+            association_methods = ::ActiveRecord::AssociationRelation.instance_methods -
+              ::ActiveRecord::Relation.instance_methods
+
+            association_methods.each do |method_name|
+              case method_name
+              when :insert_all, :insert_all!, :upsert_all
+                create_method(
+                  klass,
+                  method_name,
+                  parameters: [
+                    Parlour::RbiGenerator::Parameter.new("attributes", type: "T::Array[Hash]"),
+                    Parlour::RbiGenerator::Parameter.new("returning:", type: "T::Array[Symbol]", default: "nil"),
+                    Parlour::RbiGenerator::Parameter.new(
+                      "unique_by:",
+                      type: "T.any(T::Array[Symbol], Symbol)",
+                      default: "nil"
+                    ),
+                  ],
+                  return_type: "ActiveRecord::Result"
+                )
+              when :insert, :insert!, :upsert
+                create_method(
+                  klass,
+                  method_name,
+                  parameters: [
+                    Parlour::RbiGenerator::Parameter.new("attributes", type: "Hash"),
+                    Parlour::RbiGenerator::Parameter.new("returning:", type: "T::Array[Symbol]", default: "nil"),
+                    Parlour::RbiGenerator::Parameter.new(
+                      "unique_by:",
+                      type: "T.any(T::Array[Symbol], Symbol)",
+                      default: "nil"
+                    ),
+                  ],
+                  return_type: "ActiveRecord::Result"
+                )
+              when :proxy_association
+                # skip - private method
```

Comment:
For now, this is the only private method on that module, yeah, so that is the only one we explicitly skip (so that it doesn't look like an omission). If any more private methods are added in Rails, we won't match them anyway, since we opt-in to the methods we generate instead of opting out.